bfsscp/bfsscp.go

```diff
@@ -0,0 +1,440 @@
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	info, err := b.client.Stat(b.withPrefix(name))
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	return &bfs.MetaInfo{
+		Name:    name,
+		Size:    info.Size(),
+		ModTime: info.ModTime(),
+	}, nil
+}
+
+// Open implements bfs.Bucket.
+func (b *bucket) Open(ctx context.Context, name string) (bfs.Reader, error) {
+	file, err := b.client.Open(b.withPrefix(name))
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	info, err := file.Stat()
+	if err != nil {
+		_ = file.Close()
+		return nil, normError(err)
+	}
+
+	// Use file size to ensure limited read and avoid EOFs
+	return &reader{File: file, r: io.LimitReader(file, info.Size()), ctx: ctx}, nil
+}
+
+// Create implements bfs.Bucket.
+func (b *bucket) Create(ctx context.Context, name string, opts *bfs.WriteOptions) (bfs.Writer, error) {
+	f, err := ioutil.TempFile(b.config.TempDir, "bfs-scp")
+	if err != nil {
+		return nil, err
+	}
+
+	return &writer{
+		File:   f,
+		bucket: b,
+		ctx:    ctx,
+		name:   name,
+		opts:   opts,
+	}, nil
+}
+
+// Remove implements bfs.Bucket.
+func (b *bucket) Remove(ctx context.Context, name string) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	err := normError(b.client.Remove(b.withPrefix(name)))
+	if err != nil && err != bfs.ErrNotFound {
+		return err
+	}
+	return nil
+}
+
+// Close implements bfs.Bucket.
+func (b *bucket) Close() error {
+	return multierr.Combine(b.client.Close(), b.conn.Close())
+}
+
+// --------------------------------------------------------
+
+type writer struct {
+	*os.File
+
+	ctx    context.Context
+	bucket *bucket
+	name   string
+	opts   *bfs.WriteOptions
+
+	closeOnce sync.Once
+}
+
+func (w *writer) Discard() error {
+	err := os.ErrClosed
+	w.closeOnce.Do(func() {
+		// delete tempfile in the end
+		fname := w.Name()
+		defer os.Remove(fname)
+
+		// close tempfile
+		err = w.File.Close()
+	})
+	return err
+}
+
+func (w *writer) Commit() error {
+	err := os.ErrClosed
+	w.closeOnce.Do(func() {
+		// delete tempfile in the end
+		fname := w.Name()
+		defer os.Remove(fname)
+
+		// close tempfile, check context
+		if err = w.File.Close(); err != nil {
+			return
+		} else if err = w.ctx.Err(); err != nil {
+			return
+		}
+
+		// reopen for reading
+		var file *os.File
+		if file, err = os.Open(fname); err != nil {
+			return
+		}
+		defer file.Close()
+
+		fullName := w.bucket.withPrefix(w.name)
+		if err = w.bucket.client.MkdirAll(path.Dir(fullName)); err != nil {
+			return
+		}
+
+		var sf *sftp.File
+		if sf, err = w.bucket.client.Create(fullName); err != nil {
+			return
+		}
+		_, err = io.Copy(sf, file)
+		return
+	})
+	return err
+}
+
+// --------------------------------------------------------
+
+type reader struct {
+	*sftp.File
+	r   io.Reader
+	ctx context.Context
+}
+
+func (r *reader) Read(out []byte) (int, error) {
+	if err := r.ctx.Err(); err != nil {
```

Comment:
Hm, is it worth keeping this `type reader` wrapper just to check for ctx being cancelled?

Reply:
it's to override `Read` with `LimitedReader.Read`, admittedly `r.r.Read` isn't very easy to differentiate.